### ISSUE TITLE
Discussion: Potential improvement with git-clean (for large projects)

### ISSUE THREAD

**Issue by dakotahawkins** [git-for-windows/git#259]
The project I work on has recently switched to Git, and a few of us are using the latest builds from this repo. It's pretty impressive.

At any rate, our project is very large. We know there are better solutions, but we were forced off our old VCS so please don't make suggestions like "just don't do that, have a smaller repo" or similar. I'm sure we have an extreme case, but I think the following idea could benefit lots of smaller repositories as well...

Anyway, our repository decompresses to around 14GB, and a full (debug and release configurations) build results in something like a 60GB folder.

Sometimes, it's desirable to do a git clean (-d -f -x, for us) to remove ALL build output/temp files everywhere.

On Windows\NTFS, without a SSD, this takes forever. A really long time. I'd guess ~30 minutes for me. That's not Git, that's Windows\NTFS.

What ISN'T slow, however, is to "delete" files by moving them somewhere on the same partition, then spawning a process to delete them in the background. I did this in batch for our build server with our previous VCS and it knocked about an hour off our build times. An hour spent deleting files, imagine that!

My suggestion is that git-for-windows optionally do this to delete files instead of deleting each item asynchronously. To get the benefit, the move-to destination has to be on the same partition, and of course it has to be somewhere you don't care about. What about a temporary directory in the .git folder (is Git cool with that?).

TL;DR: On Windows/NTFS a file-move/delete-in-background can be orders of magnitude faster than a normal file deletion.

**Comment by dakotahawkins** [comment]
If this is something people would be open to, I volunteer to make the attempt.

**Comment by dscho** [comment]
Are you sure that the reason is the deletion? Or maybe something else spooky is going on, e.g. every file is opened before it is deleted? You should be able to find out by using [SysInternals' Process Monitor](https://technet.microsoft.com/en-us/library/bb896645.aspx).

**Comment by dscho** [comment]
@dakotahawkins are you still interested in working on this issue, or can I close this ticket?
